mep/books/views.py

```diff
@@ -1,25 +1,85 @@
 from dal import autocomplete
 from django.db.models import Q
-from django.views.generic import DetailView, ListView
 from django.urls import reverse
+from django.views.generic import DetailView, ListView
+from django.views.generic.edit import FormMixin
 
+from mep.books.forms import ItemSearchForm
 from mep.books.models import Item
 from mep.books.queryset import ItemSolrQuerySet
-from mep.common.views import LabeledPagesMixin, RdfViewMixin
-from mep.common.utils import absolutize_url
 from mep.common import SCHEMA_ORG
+from mep.common.utils import absolutize_url
+from mep.common.views import (AjaxTemplateMixin, FacetJSONMixin,
+                              LabeledPagesMixin, RdfViewMixin)
 
 
-class ItemList(LabeledPagesMixin, ListView, RdfViewMixin):
+class ItemList(LabeledPagesMixin, ListView, FormMixin, AjaxTemplateMixin, FacetJSONMixin, RdfViewMixin):
     '''List page for searching and browsing library items.'''
     model = Item
     template_name = 'books/item_list.html'
+    ajax_template_name = 'books/snippets/item_results.html'
     paginate_by = 100
     context_object_name = 'items'
     rdf_type = SCHEMA_ORG.SearchResultPage
 
+    form_class = ItemSearchForm
+    _form = None
+    initial = {'sort': 'title'}
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        form_data = self.request.GET.copy()
+
+        # set defaults
+        for key, val in self.initial.items():
+            form_data.setdefault(key, val)
+
+        kwargs['data'] = form_data
+        return kwargs
+
+    def get_form(self, *args, **kwargs):
+        if not self._form:
+            self._form = super().get_form(*args, **kwargs)
+        return self._form
+
+    # map form sort to solr sort
+    solr_sort = {
+        'relevance': '-score',
+        'title': 'title_s',
+    }
+
     def get_queryset(self):
-        return ItemSolrQuerySet().order_by('title')
+        # NOTE faceting on pub date currently as a placeholder; no UI use yet
+        sqs = ItemSolrQuerySet().facet_field('pub_date_i')
+        form = self.get_form()
+
+        # empty qs if not valid
+        if not form.is_valid():
+            sqs = sqs.none()
+        # otherwise apply filters, query, sort, etc.
+        else:
+            search_opts = form.cleaned_data
+            sqs = sqs.order_by(self.solr_sort[search_opts['sort']])
+
+        self.queryset = sqs
+        return sqs
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        self._form.set_choices_from_facets(
+            self.object_list.get_facets()['facet_fields'])
+        return context
+
+    def get_page_labels(self, paginator):
+        '''generate labels for pagination'''
+        form = self.get_form()
+        # if invalid, should show 'N/A'
+        if not form.is_valid():
+            return [(1, 'N/A')]
+
+        # otherwise default to numbered pages for now
+        # NOTE could implement alpha here, but tougher for titles
```

Comment:
Why is it tougher for titles? Isn't the logic exactly the same?

Reply:
(Although implementing that's isn't really part of this chore)

Reply:
When I tested it by using the same implementation as for names, the results were _really_ long (like > 100 chars in many cases), so I assumed we would have to do something special if we wanted it to work.

Reply:
Ah, good to know. I can imagine truncating it at some max length (which we might want to do for names also).  And I think we don't yet have sort titles; that might help some.

I'm curious now if it would be useful for some of the other sorts, like publication date...